src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -99,94 +159,96 @@ internal static string IPv6AddressToString(ushort[] address, uint scopeId)
+            return buffer;
         }
 
-        private static unsafe void FormatIPv4AddressNumber(int number, char* addressString, ref int offset)
+        private static void FormatIPv4AddressNumber(int number, Span<char> addressString, ref int offset)
         {
+            // Math.DivRem has no overload for byte, assert here for safety
+            Debug.Assert(number < 256);
+
+            offset += number > 99 ? 3 : number > 9 ? 2 : 1;
+
             int i = offset;
             do
             {
-                int rem;
-                number = Math.DivRem(number, 10, out rem);
+                number = Math.DivRem(number, 10, out int rem);
                 addressString[--i] = (char)('0' + rem);
             } while (number != 0);
-            offset = i;
         }
 
-        public static unsafe bool Ipv4StringToAddress(string ipString, out long address)
+        public static unsafe bool Ipv4StringToAddress(ReadOnlySpan<char> ipSpan, out long address)
         {
-            Debug.Assert(ipString != null);
-
-            long tmpAddr;
-            int end = ipString.Length;
-            fixed (char* ipStringPtr = ipString)
+            int end = ipSpan.Length;
+            fixed (char* ipStringPtr = &ipSpan.DangerousGetPinnableReference())
             {
-                tmpAddr = IPv4AddressHelper.ParseNonCanonical(ipStringPtr, 0, ref end, notImplicitFile: true);
-            }
+                long tmpAddr = IPv4AddressHelper.ParseNonCanonical(ipStringPtr, 0, ref end, notImplicitFile: true);
 
-            if (tmpAddr != IPv4AddressHelper.Invalid && end == ipString.Length)
-            {
-                // IPv4AddressHelper.ParseNonCanonical returns the bytes in the inverse order.
-                // Reverse them and return success.
-                address =
-                    ((0xFF000000 & tmpAddr) >> 24) |
-                    ((0x00FF0000 & tmpAddr) >> 8) |
-                    ((0x0000FF00 & tmpAddr) << 8) |
-                    ((0x000000FF & tmpAddr) << 24);
-                return true;
-            }
-            else
-            {
-                // Failed to parse the address.
-                address = 0;
-                return false;
+                if (tmpAddr != IPv4AddressHelper.Invalid && end == ipSpan.Length)
+                {
+                    // IPv4AddressHelper.ParseNonCanonical returns the bytes in the inverse order.
+                    // Reverse them and return success.
+                    address =
+                        ((0xFF000000 & tmpAddr) >> 24) |
+                        ((0x00FF0000 & tmpAddr) >> 8) |
+                        ((0x0000FF00 & tmpAddr) << 8) |
+                        ((0x000000FF & tmpAddr) << 24);
+                    return true;
+                }
+                else
+                {
+                    // Failed to parse the address.
+                    address = 0;
+                    return false;
+                }
```

Comment:
Does the pinning need to extend down to here?  Seems like it only need to be done just for the ParseNonCanonical call, as it was in the old code.